src/approaches/grammar_search_invention_approach.py

```diff
@@ -396,6 +397,220 @@ def _create_grammar(grammar_name: str, dataset: Dataset,
+            hadd_fn = utils.HAddHeuristic(
+                utils.atoms_to_tuples(init_atoms),
+                utils.atoms_to_tuples(goal_atoms),
+                relaxed_operators)
+            score += self._evaluate_atom_trajectory(atoms_sequence, hadd_fn,
+                                                    ground_ops)
+        return CFG.grammar_search_lookahead_hadd_weight * score
+
+    def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
+                                  hadd_fn: utils.HAddHeuristic,
+                                  ground_ops: Collection[_GroundSTRIPSOperator]
+                                  ) -> float:
+        raise NotImplementedError("Override me!")
+
+
+class _HAddMatchHeuristic(_HAddBasedHeuristic):
+    """Score based on distance to "groundtruth" value function for actions that
+    were seen in the demonstration.
+    """
+
+    def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
+                                  hadd_fn: utils.HAddHeuristic,
+                                  ground_ops: Collection[_GroundSTRIPSOperator]
+                                  ) -> float:
+        score = 0.0
+        for i, atoms in enumerate(atoms_sequence):
+            ideal_heuristic = len(atoms_sequence) - i - 1
+            hadd_heuristic = hadd_fn(utils.atoms_to_tuples(atoms))
+            score += abs(hadd_heuristic - ideal_heuristic)
+        return score
+
+
+class _HaddLookaheadHeuristic(_HAddBasedHeuristic):
+    """Score predicates by using the hadd values of the induced operators
+    to compute an energy-based policy, and comparing that policy to demos.
+
+    Overview of the idea:
+    1. Predicates induce operators. Denote this ops(preds).
+    2. Operators induce an hadd heuristic. Denote this hadd(state, ops(preds)).
+    3. The heuristic induces a greedy one-step lookahead energy-based policy.
+       Denote this pi(a|s) propto exp(-constant * hadd(succ(s, a), ops(preds)).
+    4. The objective for predicate learning is to maximize prod pi(a | s)
+       where the product is over demonstrations.
+    """
+
+    def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
+                                  hadd_fn: utils.HAddHeuristic,
+                                  ground_ops: Collection[_GroundSTRIPSOperator]
+                                  ) -> float:
+        score = 0
+        for i in range(len(atoms_sequence)-1):
+            atoms, next_atoms = atoms_sequence[i], atoms_sequence[i+1]
+            ground_op_demo_prob = 0. # total probability mass for demo actions
+            ground_op_total_prob = 0.  # total probability mass for all actions
+            for ground_op in ground_ops:
+                # Only care about applicable ground ops.
+                if not ground_op.preconditions.issubset(atoms):
+                    continue
+                # Compute the next state under the operator.
```

Comment:
We have a thing called `apply_nsrt` in utils.py, we should also make a `apply_operator` and use that here. And copy the tests for `apply_nsrt` to make some tests for `apply_operator` as well.